File: reference/internal/ui/components/commentform.go
package components

import (
	"strings"

	"github.com/Elpulgo/azdo/internal/ui/styles"
	"github.com/charmbracelet/bubbles/textarea"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// CommentSubmittedMsg is emitted when the user submits a comment (Ctrl+S).
type CommentSubmittedMsg struct {
	Text string
}

// CommentFormCancelledMsg is emitted when the user cancels the form (Esc).
type CommentFormCancelledMsg struct{}

// commentFormHeight is the number of textarea rows shown in the form.
const commentFormHeight = 5

// CommentForm is an inline multi-line text form for composing a work item comment.
// Enter inserts a newline; Ctrl+S submits; Esc cancels.
type CommentForm struct {
	styles   *styles.Styles
	textarea textarea.Model
	visible  bool
}

// NewCommentForm creates a new inline comment form.
func NewCommentForm(s *styles.Styles) CommentForm {
	ta := textarea.New()
	ta.Placeholder = "Write a comment..."
	ta.ShowLineNumbers = false
	ta.CharLimit = 0 // no limit
	ta.SetHeight(commentFormHeight)

	return CommentForm{
		styles:   s,
		textarea: ta,
		visible:  false,
	}
}

// Show makes the form visible.
func (f *CommentForm) Show() {
	f.visible = true
}

// Hide makes the form invisible and blurs the textarea.
func (f *CommentForm) Hide() {
	f.visible = false
	f.textarea.Blur()
}

// IsVisible returns whether the form is visible.
func (f CommentForm) IsVisible() bool {
	return f.visible
}

// Focus focuses the textarea so it captures keystrokes.
func (f *CommentForm) Focus() tea.Cmd {
	return f.textarea.Focus()
}

// Reset clears the textarea content.
func (f *CommentForm) Reset() {
	f.textarea.Reset()
}

// SetValue replaces the textarea content (used to restore a draft after a failed send).
func (f *CommentForm) SetValue(s string) {
	f.textarea.SetValue(s)
}

// Value returns the current textarea content.
func (f CommentForm) Value() string {
	return f.textarea.Value()
}

// SetWidth sizes the textarea to the available width.
func (f *CommentForm) SetWidth(width int) {
	// Leave a little room for the border/padding.
	w := width - 4
	if w < 10 {
		w = 10
	}
	f.textarea.SetWidth(w)
}

// Height returns the number of terminal rows the form occupies when visible
// (textarea rows + border + help line).
func (f CommentForm) Height() int {
	// textarea height + top/bottom border (2) + help line (1)
	return commentFormHeight + 3
}

// Update handles messages for the form. Ctrl+S submits, Esc cancels, everything
// else (including Enter) is delegated to the textarea.
func (f CommentForm) Update(msg tea.Msg) (CommentForm, tea.Cmd) {
	if !f.visible {
		return f, nil
	}

	if key, ok := msg.(tea.KeyMsg); ok {
		switch key.Type {
		case tea.KeyEsc:
			// Hide synchronously so the dispatched CommentFormCancelledMsg is
			// handled by the parent rather than re-captured by this form.
			f.visible = false
			f.textarea.Blur()
			return f, func() tea.Msg { return CommentFormCancelledMsg{} }
		case tea.KeyCtrlS:
			text := f.textarea.Value()
			if strings.TrimSpace(text) == "" {
				return f, nil
			}
			f.visible = false
			f.textarea.Blur()
			return f, func() tea.Msg { return CommentSubmittedMsg{Text: text} }
		}
	}

	var cmd tea.Cmd
	f.textarea, cmd = f.textarea.Update(msg)
	return f, cmd
}

// View renders the inline form.
func (f CommentForm) View() string {
	if !f.visible {
		return ""
	}

	helpText := lipgloss.NewStyle().
		Foreground(f.styles.Theme.GetForegroundMuted()).
		Render("Ctrl+S: send • Esc: cancel")

	box := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(f.styles.Theme.GetBorder()).
		Render(f.textarea.View())

	return lipgloss.JoinVertical(lipgloss.Left, box, helpText)
}

File: src/Azdo.Tui/Components/CommentForm.cs
using System.Text;
using Azdo.Tui.Rendering;
using Azdo.Tui.Runtime;
using StyleSet = Azdo.Tui.Styles.Styles;

namespace Azdo.Tui.Components;

/// <summary>Emitted when the user submits a comment with Ctrl+S (≈ <c>components.CommentSubmittedMsg</c>).</summary>
public sealed record CommentSubmittedMsg(string Text) : IMsg;

/// <summary>Emitted when the user cancels the form with Esc (≈ <c>components.CommentFormCancelledMsg</c>).</summary>
public sealed record CommentFormCancelledMsg : IMsg
{
    public static readonly CommentFormCancelledMsg Instance = new();
}

/// <summary>
/// An inline multi-line text form for composing a work item comment
/// (≈ <c>components.CommentForm</c>). Enter inserts a newline; Ctrl+S submits;
/// Esc cancels. A minimal textarea is implemented inline since the foundation
/// <see cref="TextInput"/> is single-line.
/// </summary>
public sealed class CommentForm
{
    private const int CommentFormHeight = 5; // textarea rows shown
    private const string Placeholder = "Write a comment...";

    private readonly StyleSet _styles;
    private readonly StringBuilder _value = new();
    private bool _visible;
    private bool _focused;
    private int _width = 40;

    public CommentForm(StyleSet styles) => _styles = styles;

    public void Show() => _visible = true;
    public void Hide() { _visible = false; _focused = false; }
    public bool IsVisible => _visible;

    /// <summary>Focuses the textarea so it captures keystrokes. Returns no command.</summary>
    public Cmd? Focus() { _focused = true; return null; }

    public void Reset() => _value.Clear();
    public void SetValue(string s) { _value.Clear(); _value.Append(s); }
    public string Value() => _value.ToString();

    /// <summary>Sizes the textarea to the available width (leaving room for border/padding).</summary>
    public void SetWidth(int width)
    {
        int w = width - 4;
        if (w < 10) w = 10;
        _width = w;
    }

    /// <summary>Terminal rows the form occupies when visible (textarea + border + help line).</summary>
    public int Height() => CommentFormHeight + 3;

    /// <summary>
    /// Handles a message. Ctrl+S submits, Esc cancels, Enter inserts a newline,
    /// everything else is delegated to the inline textarea.
    /// </summary>
    public Cmd? Update(IMsg msg)
    {
        if (!_visible) return null;

        if (msg is KeyMsg key)
        {
            switch (key.Key)
            {
                case "esc":
                    // Hide synchronously so the dispatched cancel message is handled by the parent.
                    _visible = false;
                    _focused = false;
                    return Commands.Of(CommentFormCancelledMsg.Instance);

                case "ctrl+s":
                    string text = _value.ToString();
                    if (string.IsNullOrWhiteSpace(text)) return null;
                    _visible = false;
                    _focused = false;
                    return Commands.Of(new CommentSubmittedMsg(text));

                case "enter":
                    _value.Append('\n');
                    return null;

                case "backspace":
                    if (_value.Length > 0) _value.Remove(_value.Length - 1, 1);
                    return null;

                case " ":
                    _value.Append(' ');
                    return null;

                default:
                    if (key.IsRune) _value.Append(key.Rune);
                    return null;
            }
        }

        return null;
    }

    public string View()
    {
        if (!_visible) return "";

        var theme = _styles.Theme;

        string helpText = Style.New()
            .Foreground(theme.ForegroundMuted)
            .Render("Ctrl+S: send • Esc: cancel");

        string textareaView = RenderTextarea();

        string box = Style.New()
            .WithBorder(Border.Rounded)
            .BorderForeground(theme.Border)
            .Render(textareaView);

        return Layout.JoinVertical(HAlign.Left, box, helpText);
    }

    /// <summary>Renders the textarea body: fixed height, padded to the configured width.</summary>
    private string RenderTextarea()
    {
        string content = _value.Length == 0 && !_focused
            ? Placeholder
            : _value.ToString();

        var lines = content.Split('\n').ToList();
        while (lines.Count < CommentFormHeight) lines.Add("");
        if (lines.Count > CommentFormHeight)
            lines = lines.GetRange(lines.Count - CommentFormHeight, CommentFormHeight);

        for (int i = 0; i < lines.Count; i++)
            lines[i] = Ansi.PadRight(Ansi.Truncate(lines[i], _width), _width);

        return string.Join("\n", lines);
    }
}
